csrc/ir/utils.cpp

```diff
@@ -1006,6 +1005,78 @@ class ValidateDomainEquivalence : private IterVisitor {
   std::unordered_set<Val*> frontier_;
 };
 
+std::vector<Statement*> next(Statement* stmt) {
+  if (stmt->isVal()) {
+    if (auto val = stmt->as<Val>()->definition()) {
+      return {val};
+    } else {
+      return {};
+    }
+  } else {
+    auto expr = stmt->as<Expr>();
+    std::vector<Statement*> inputs{
+        expr->inputs().begin(), expr->inputs().end()};
+    return inputs;
+  }
+}
+
+std::vector<Statement*> checkCycle(
```

Comment:
Can you please make this also look at sibling nodes no matter if they are visited? For example, a WelfordOp produces three tensors, and suppose only one is set as a fusion output, but the other two tensors are still generated, so we generally traverse them as well (e.g., `Fusion::usedMathVals()` includes such tensors). If we accidentally modified them, they may end up resulting in a cycle.

Reply:
That's a fun thing to know. I thought we left LOTS of dead code in the fusion definition, so I was using the `to` vector here to limit the search scope.

Since the traversal we have traces through producers. If we want to traverse siblings, we need to go through consumers instead. Which won't give us the opportunity to filter out section with `to`.
We can patch it by filtering out all ops with more than one output and traverse the other direction to figure out if there's any loop there. Which unfortunately is going to look messy in the implementation, since the signature here suggests that we are only checking sections between `to` to `from`, but that's not the case any more. :cry:

Reply:
Maybe we could just handle them after the cycle check with the used outputs? Once the initial check is done, we know which nodes are siblings that haven't showed up. We could do the same cycle check starting with those sibling nodes. Isn't it enough?

Reply:
I think that's a nice middle ground to solve it in the short-term. I'll add that logic here.

> We could do the same cycle check starting with those sibling nodes

Just want to point it out that, we just need to do the cyclic check on the other directions though, following the consumers instead. Since we don't know the `end node` on the sibling branch. But that's not a big deal.

Meanwhile, back to the topic on dangling statements in the fusion IR that's contributing to codegen. This doesn't sound right to me and I think we need to also working on a coherent way to handle that. The logic in `usedMathVals` feels much like a hack...

Reply:
This is actually not a concern as any cycle that would involve dangling siblings should be either detected by the traversal from non-dangling consumers or should just be ignored. Thanks @jjsjann123